lib/compress.js

```diff
@@ -620,20 +621,21 @@ merge(Compressor.prototype, {
                                 || (parent instanceof AST_If          && node !== parent.condition)
                                 || (parent instanceof AST_Conditional && node !== parent.condition)
                                 || (node instanceof AST_SymbolRef
-                                    && !are_references_in_scope(node.definition(), self))
+                                    && value_has_side_effects
+                                    && !are_references_in_scope(node.definition()))
                                 || (parent instanceof AST_Binary
                                     && (parent.operator == "&&" || parent.operator == "||")
                                     && node === parent.right)
                                 || (parent instanceof AST_Switch && node !== parent.expression)) {
                                 return side_effects_encountered = unwind = true, node;
                             }
-                            function are_references_in_scope(def, scope) {
+                            function are_references_in_scope(def) {
```

Comment:
Just feels odd seeing an inner function not using available inner variables.

But alright, I'll revert those.